### PR TITLE
add dotnet new templates (razor view, api controller, mvc controller)

### PR DESCRIPTION
**Pull request by deepchoudhery**
added new `dotnet new` web item templates
Razor View:
- added `view`
- adds an empty razor view (Index.cshtml) to the project.
- same as `dotnet new page -np` I guess so disregard this for now
- [Help](https://microsoft-my.sharepoint.com/:i:/p/decho/EfPlHAfyNTBIodaHN-UdC_kBhVxYyltmIGoliCXcgOX--w?e=4w5jmr)

API Controller:
- added `apiController`
- adds an empty API Controller.
- use `--actions` to add one with read/write actions.
- [Help](https://microsoft-my.sharepoint.com/:i:/p/decho/EYxbSRPmfgVDuDL55WgTY4MBKTKWl2vpAmPiSxvleuO7yw?e=pdbDqX)

MVC Controller:
- added `mvcController`
- adds an empty MVC Controller.
- use `--actions` to add one with read/write actions.
- [Help](https://microsoft-my.sharepoint.com/:i:/p/decho/EYxbSRPmfgVDuDL55WgTY4MBKTKWl2vpAmPiSxvleuO7yw?e=pdbDqX)

### CODE REVIEW

File: src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/template.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  "name": "Razor View",
+  "generatorVersions": "[1.0.0.0-*)",
+  "description": "Am empty razor view",
+  "tags": {
+    "language": "C#",
+    "type": "item"
+  },
+  "groupIdentity": "Microsoft.AspNetCore.Mvc.RazorView",
+  "precedence": "9801",
+  "identity": "Microsoft.AspNetCore.Mvc.RazorView.7.0",
```
Comment (vijayrkn) [inline src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/template.json]: shouldn't this be 8.0 to match the other templates in main?

File: src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/localize/templatestrings.en.json [hunk 2]
```diff
@@ -0,0 +1,6 @@
+{
+  "author": "Microsoft",
+  "name": "Razor View",
+  "description": "A Razor view without a page model",
```
Comment (vijayrkn) [inline src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/localize/templatestrings.en.json:4]: Let's review these strings with Sayed as well.

**Review by vijayrkn — CHANGES_REQUESTED** [review]

File: src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/template.json (hunk 1 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/template.json]: fixed.

File: src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/localize/templatestrings.en.json (hunk 2 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/localize/templatestrings.en.json:4]: yup this is more of a POC, won't be actually merging this PR, just a quick sample

**Comment by vijayrkn** [comment]
@Phenning - Can you please help take a look if anything stands out? These are pretty basic item templates.

**Comment by phenning** [comment]
@vijayrkn is there a specific problem, or just a general review?

@deepchoudhery I'm assuming we aren't expecting these to show up in VS right away (they won't as authored), as we need to solve the problem on how to suppress the current vstemplate versions of these templates.

File: src/ProjectTemplates/Web.ItemTemplates/content/ApiController/ValueController.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
+using Microsoft.AspNetCore.Http;
```
Comment (vijayrkn) [inline src/ProjectTemplates/Web.ItemTemplates/content/ApiController/ValueController.cs:1]: Looks so much cleaner than the cshtml equivalents.

**Comment by vijayrkn** [comment]
> @vijayrkn is there a specific problem, or just a general review?

Just a general review.

We are not planning to ship these in VS. This is to move the templates out of the scaffolding repo. VS is already relying on the item templates that are shipped in VS.

File: src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  "classifications": [ "Web", "ASP.NET" ],
+  "name": "API Controller",
+  "generatorVersions": "[1.0.0.0-*)",
+  "description": "An API Controller with or without read/write actions",
+  "tags": {
+    "language": "C#",
+    "type": "item"
+  },
+  "groupIdentity": "Microsoft.AspNetCore.Mvc.ApiController",
+  "precedence": "9799",
```
Comment (phenning) [inline src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json]: Any reason why 9799 and not like 9800 like the other ones in dotnet/aspnetcore?

File: src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  "defaultName": "ValueController",
+  "symbols": {
+    "namespace": {
+      "description": "namespace for the generated code",
+      "replaces": "MyApp.Namespace",
+      "type": "parameter"
+    },
+    "actions": {
+      "description": "create controller with read/write actions",
+      "displayName": "add read/write actions",
```
Comment (phenning) [inline src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json]: displayName is only used by Visual Studio, and would display exactly this way, casing and all. I think the displayName should be sentence or pascal cased (See Razor Page template)

However, in reality the item dialog as implemented does not support showing parameters anyway.

File: src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/localize/templatestrings.en.json [hunk 6]
```diff
@@ -0,0 +1,9 @@
+{
+  "author": "Microsoft",
+  "name": "API Controller",
+  "description": "An API Controller with or without read/write actions",
+  "symbols/namespace/description": "namespace for the generated code",
+  "symbols/actions/description": "create controller with read/write actions",
+  "symbols/actions/displayName": "Include read/write actions",
```
Comment (phenning) [inline src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/localize/templatestrings.en.json]: Inconsistent with template.json string, see comment there.

File: src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  "defaultName": "HomeController",
+  "symbols": {
+    "namespace": {
+      "description": "namespace for the generated code",
+      "replaces": "MyApp.Namespace",
+      "type": "parameter"
+    },
+    "actions": {
+      "description": "create controller with read/write actions",
+      "displayName": "add read/write actions",
```
Comment (phenning) [inline src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json]: Same comment as APIController

File: src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/localize/templatestrings.en.json [hunk 8]
```diff
@@ -0,0 +1,9 @@
+{
+  "author": "Microsoft",
+  "name": "MVC Controller",
+  "description": "A MVC Controller with or without read/write actions",
+  "symbols/namespace/description": "namespace for the generated code",
+  "symbols/actions/description": "create controller with read/write actions",
+  "symbols/actions/displayName": "Include read/write actions",
```
Comment (phenning) [inline src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/localize/templatestrings.en.json]: See APIController.

File: src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  "classifications": [ "Web", "ASP.NET" ],
+  "name": "MVC Controller",
+  "generatorVersions": "[1.0.0.0-*)",
+  "description": "A MVC Controller with or without read/write actions",
+  "tags": {
+    "language": "C#",
+    "type": "item"
+  },
+  "groupIdentity": "Microsoft.AspNetCore.Mvc.MvcController",
+  "precedence": "9799",
```
Comment (phenning) [inline src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json]: 9800?

File: src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/template.json [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  "classifications": [ "Web", "ASP.NET" ],
+  "name": "Razor View",
+  "generatorVersions": "[1.0.0.0-*)",
+  "description": "Am empty razor view",
+  "tags": {
+    "language": "C#",
+    "type": "item"
+  },
+  "groupIdentity": "Microsoft.AspNetCore.Mvc.RazorView",
+  "precedence": "9801",
```
Comment (phenning) [inline src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/template.json]: why 9801?

**Comment by phenning** [comment]
> > @vijayrkn is there a specific problem, or just a general review?
> 
> Just a general review.
> 
> We are not planning to ship these in VS. This is to move the templates out of the scaffolding repo. VS is already relying on the item templates that are shipped in VS.

Will these ship in the SDK?  I know we wouldn't have them dynamically show up in VS when they do, due to missing metadata to enable it.  If we would want to invoke them from vs, we could probably figure out a way to invoke them from the vstemplate stubs instead of the content in the vstemplate. Would probably need a custom wizard, or modification of our existing vstemplate/template engine wizard to support the scenario.

File: src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json (hunk 4 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json]: I think I was trying to give it an order, this I was hoping a PM would have an opinion on. We have a meeting about this I'll add you to if you have time thursday

Comment (phenning) [reply src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json]: see [this PR](https://github.com/dotnet/aspnetcore/pull/45979#discussion_r1065265430) where doug/william settled on 9800 for 8.0

File: src/ProjectTemplates/Web.ItemTemplates/content/ApiController/ValueController.cs (hunk 3 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/ApiController/ValueController.cs:1]: yes it does

File: src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json (hunk 5 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json]: fixed

File: src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json (hunk 4 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json]: changing all to 9800 for now.

File: src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json (hunk 9 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json]: changed

File: src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/template.json (hunk 10 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/template.json]: changed to 9800.

File: src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json (hunk 7 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json]: edited.

File: src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/localize/templatestrings.en.json (hunk 6 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/localize/templatestrings.en.json]: fixed ty

File: src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/localize/templatestrings.en.json (hunk 8 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/localize/templatestrings.en.json]: fixed ty

**Comment by vijayrkn** [comment]
> If we would want to invoke them from vs, we could probably figure out a way to invoke them from the vstemplate stubs instead of the content in the vstemplate.

For now, we have vstemplates for all of these item templates. But we can definitely consider moving to these in the future to have 1 source of truth.

**Comment by sayedihashimi** [comment]
The names `apiController` and `mvcController` should be all lower case, that is how all the other templates are defined. The other alternative is to use `controller` and then have a switch for `mvc` or `api` but I don't know which one of those should be the default. Maybe it's better to keep them separate like you have now.

**Comment by sayedihashimi** [comment]
I think we can delete the `dotnetcli.host.json` file since it only has a schema declaration. Right @phenning?

File: src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/localize/templatestrings.en.json [hunk 11]
```diff
@@ -0,0 +1,9 @@
+{
+  "author": "Microsoft",
+  "name": "API Controller",
+  "description": "An API Controller with or without read/write actions",
```
Comment (sayedihashimi) [inline src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/localize/templatestrings.en.json]: We can remove the "an" at the beginning. `API Controller with or without read/write actions`

File: src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json [hunk 12]
```diff
@@ -0,0 +1,58 @@
+{
+  "$schema": "http://json.schemastore.org/template",
+  "author": "Microsoft",
+  "classifications": [ "Web", "ASP.NET" ],
+  "name": "API Controller",
+  "generatorVersions": "[1.0.0.0-*)",
+  "description": "An API Controller with or without read/write actions",
```
Comment (sayedihashimi) [inline src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json]: We can drop the "an" `API Controller with or without read/write actions`

File: src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/localize/templatestrings.en.json [hunk 13]
```diff
@@ -0,0 +1,9 @@
+{
+  "author": "Microsoft",
+  "name": "MVC Controller",
+  "description": "A MVC Controller with or without read/write actions",
```
Comment (sayedihashimi) [inline src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/localize/templatestrings.en.json]: Drop the leading "A" to make it `MVC Controller with or without read/write actions`

File: src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json [hunk 14]
```diff
@@ -0,0 +1,58 @@
+{
+  "$schema": "http://json.schemastore.org/template",
+  "author": "Microsoft",
+  "classifications": [ "Web", "ASP.NET" ],
+  "name": "MVC Controller",
+  "generatorVersions": "[1.0.0.0-*)",
+  "description": "A MVC Controller with or without read/write actions",
```
Comment (sayedihashimi) [inline src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json]: Drop the leading "a" to make it `MVC Controller with or without read/write actions`

File: src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+  "tags": {
+    "language": "C#",
+    "type": "item"
+  },
+  "groupIdentity": "Microsoft.AspNetCore.Mvc.MvcController",
+  "precedence": "9800",
+  "identity": "Microsoft.AspNetCore.Mvc.MvcController.8.0",
+  "shortName": "mvcController",
```
Comment (sayedihashimi) [inline src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json]: Make all lowercase.

File: src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+  "tags": {
+    "language": "C#",
+    "type": "item"
+  },
+  "groupIdentity": "Microsoft.AspNetCore.Mvc.ApiController",
+  "precedence": "9800",
+  "identity": "Microsoft.AspNetCore.Mvc.ApiController.8.0",
+  "shortName": "apiController",
```
Comment (sayedihashimi) [inline src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json]: Make all lowercase.

File: src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+      "displayName": "Add ReadWrite Actions",
+      "type": "parameter",
+      "datatype": "bool",
+      "defaultValue": "false"
+    },
+    "HostIdentifier": {
+      "type": "bind",
+      "binding": "HostIdentifier"
+    },
+    "NameIsController": {
```
Comment (sayedihashimi) [inline src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json:40]: I don't see this getting used in the HomeController.cs file.

File: src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+      "type": "computed",
+      "value": "(name == \"Controller\")"
+    }
+  },
+  "postActions": [
+    {
+      "id": "openInEditor",
+      "condition": "(HostIdentifier != \"dotnetcli\" && HostIdentifier != \"dotnetcli-preview\")",
+      "description": "Opens HomeController.cs in the editor",
```
Comment (sayedihashimi) [inline src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json:49]: If I'm not mistaken the user will only see the name `HomeController` if they name it as such. Seems like an implementation detail that the user is not aware of. Should we change this to "Opens the created controller in the editor"?

cc @phenning

File: src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/localize/templatestrings.en.json [hunk 19]
```diff
@@ -0,0 +1,6 @@
...
+  "postActions/openInEditor/description": "Opens Index.cshtml in the editor"
```
Comment (sayedihashimi) [inline src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/localize/templatestrings.en.json:5]: Similar comment regarding `Index.cshtml` as to the HomeController one. Maybe it's better to state "Opens the generated view in the editor"

File: src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/template.json [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+  "generatorVersions": "[1.0.0.0-*)",
+  "description": "Am empty razor view",
+  "tags": {
+    "language": "C#",
+    "type": "item"
+  },
+  "groupIdentity": "Microsoft.AspNetCore.Mvc.RazorView",
+  "precedence": "9800",
+  "identity": "Microsoft.AspNetCore.Mvc.RazorView.8.0",
+  "shortName": "view",
```
Comment (sayedihashimi) [inline src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/template.json:15]: `view` is pretty generic, should this be `razorview`?

cc @phenning

File: src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/template.json [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+    "HostIdentifier": {
+      "type": "bind",
+      "binding": "HostIdentifier"
+    }
+  },
+  "postActions": [
+    {
+      "id": "openInEditor",
+      "condition": "(HostIdentifier != \"dotnetcli\" && HostIdentifier != \"dotnetcli-preview\")",
+      "description": "Opens Index.cshtml in the editor",
```
Comment (sayedihashimi) [inline src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/template.json:31]: See previous comment regarding this string.

**Review by sayedihashimi — CHANGES_REQUESTED** [review]
I have left some comments that we should review.

File: src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/localize/templatestrings.en.json (hunk 11 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/localize/templatestrings.en.json]: sounds good ty

File: src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/localize/templatestrings.en.json (hunk 13 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/localize/templatestrings.en.json]: sg ty

File: src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json (hunk 12 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/ApiController/.template.config/template.json]: sgty

File: src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json (hunk 15 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json]: sgsg

File: src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json (hunk 17 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json:40]: missed it, using it now.

File: src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json (hunk 18 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/MvcController/.template.config/template.json:49]: sure yeah will change,

File: src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/template.json (hunk 21 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/template.json:31]: fixed ty

File: src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/localize/templatestrings.en.json (hunk 19 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/localize/templatestrings.en.json:5]: kept it `Opens created view in the editor` to keep it consistent with `Opens created controller in the editor` for the controller scenarios.

Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/localize/templatestrings.en.json:5]: looks good @sayedihashimi ?

Comment (sayedihashimi) [reply src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/localize/templatestrings.en.json:5]: "Opens created view in the editor" sounds good, thanks.

File: src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/template.json (hunk 20 above)
Comment (deepchoudhery) [reply src/ProjectTemplates/Web.ItemTemplates/content/RazorView/.template.config/template.json:15]: will discuss this in the PR into aspnetcore

**Pull request merged**